add IonQSampler and IonQEstimator classes

splch opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short, detailed and understandable for all.
⚠️ If your pull request addresses an open issue, please link to the issue.

✅ I have added tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.

😊 Thank you for helping make this project better!
-->

### Summary

adds support for the sampler and estimator primitives. resolves issue #242 

### Details and comments

the sampler expects `memory=True`, so this will need to be updated after #234 is merged
also added testing for python 3.14 🐍🥧

splch committed:
add IonQSampler and IonQEstimator classes

splch committed:
update header comments and license information in ionq_primitives.py

hdnq commented on qiskit_ionq/ionq_primitives.py:
```diff
@@ -0,0 +1,164 @@
+        # Normalize everything to SamplerPub objects with shots filled in.
+        coerced_pubs = [SamplerPub.coerce(pub, shots) for pub in pubs]
+
+        job = PrimitiveJob(self._run, coerced_pubs)
```
@splch I think this is the origin of one of the user-reported bugs, which is not fixed here - qiskit's BackendSamplerV2 creates and returns a PrimitiveJob (with random uuid) to the user, and the same thing is still happening here. There's an IonQJob created in `_run_pub` but only the result, not the job, gets passed back up.

so when a user does this
```
job = sampler.run([qc], shots=100)
job.job_id()
```
they'll still just get a random uuid back and not the uuid for the actual job in our platform. They want the actual uuid for the job so they can store it, retrieve the result later, etc. (I think Estimator does the same thing)

splch replied:
thanks for the reminder! i'll get that added

splch committed:
update copyright information and add tests for IonQSampler and IonQEstimator

splch committed:
ignore SparseEfficiencyWarning in pytest configuration

splch committed:
enhance IonQSampler to include job IDs in results metadata

splch committed:
remove unused EstimatorPub containers from ionq_primitives.py

splch committed:
add IonQ job ID tracking in IonQSampler and IonQEstimator metadata

splch committed:
update Python version to 3.14 in CI workflows, versioning, and tox configuration

splch committed:
Surface IonQ job IDs via BasePrimitiveJob (closes #242)

The prior shape returned a vanilla `qiskit.primitives.PrimitiveJob` from
`IonQSampler.run()` / `IonQEstimator.run()`. That class hardcodes
`str(uuid.uuid4())` at construction (qiskit/primitives/primitive_job.py:39),
so `backend.retrieve_job(sampler_job.job_id())` returned 404 - the exact
bug reported in #242. Surfacing IonQ IDs only via `result.metadata` was a
workaround, not a fix: users persist `job.job_id()` per the documented
Qiskit API.

Rewrite to mirror the qiskit-braket-provider v0.11 / qiskit-ibm-runtime
`RuntimeJobV2` pattern (the only shipping third-party V2-primitive
precedents):

- New `IonQPrimitiveJob(BasePrimitiveJob)` whose `.job_id()` returns the
  IonQ-server-assigned UUID. `PrimitiveJob`'s own docstring at
  primitives/primitive_job.py:28-32 explicitly tells third-party
  implementers to subclass `BasePrimitiveJob` instead.
- Multi-PUB convention follows IBM Runtime: one `.job_id()` per
  `run()` call (the first IonQ UUID); full list via the
  `.ionq_job_ids` accessor and `result.metadata['ionq_job_ids']`. Per-PUB
  IDs are also stamped on each `SamplerPubResult` / `PubResult` metadata.
- `IonQSampler` now subclasses `BaseSamplerV2` directly and owns the
  counts -> `BitArray` path, so it no longer depends on
  `BackendSamplerV2`'s hardcoded `memory=True` shotwise contract. Eager
  submission at `run()` time means the IonQ UUIDs are known immediately.
- `IonQEstimator` keeps `BackendEstimatorV2` inheritance (reuses the
  observable-grouping / Pauli-basis / expectation-value helpers) but
  overrides `run()` to also return `IonQPrimitiveJob`. A pure
  `BaseEstimatorV2` subclass for full symmetry is deferred - it would
  require re-vendoring ~150 lines of private estimator helpers and is
  not necessary to close #242.
- Server-side discoverability: every primitive submission is
  auto-tagged with `extra_metadata={qiskit_primitive: sampler|estimator,
  qiskit_primitive_pub_index: i, qiskit_primitive_pub_count: n}` so jobs
  remain findable in the IonQ Cloud UI even when only the Qiskit-side
  `job_id` is lost. User-supplied `extra_metadata` wins on collision.
- Fix: `_counts_to_bitarrays` now sums when joined bitstrings collapse
  to the same per-register key (e.g. '00' and '10' both -> c[0]='0')
  instead of overwriting. Regression-guarded by a new test.

25 tests pin the contract: `.job_id()` returns the IonQ UUID (not a
random UUID), the `BasePrimitiveJob` abstract surface (status
aggregation, cancel fan-out, lazy result memoization), multi-PUB ID
exposure, auto-tagging precedence, and per-register `BitArray`
construction. Full suite: 364 passed.

splch committed:
Merge main into estimator-and-sampler

Brings the PR up to date with main (8 commits behind since Dec 5, 2025).
Almost all files auto-merged cleanly because the original PR #237 commits
only touched the new `ionq_primitives.py` + its test file plus a handful
of CI/version/tox knobs - main's evolution of `ionq_job.py`,
`ionq_client.py`, etc. was orthogonal.

Two real conflicts resolved:

* qiskit_ionq/version.py - took main's `1.0.3`. The PR branch had a
  speculative bump to `1.1.0` from one of the original (Dec 2025)
  primitive commits, but release versioning belongs in a separate
  release-prep commit; claiming a minor version this PR hasn't shipped
  under would conflict with whatever the next release actually is.

* tox.ini - took main's 2-space pytest filter indentation and main's
  new `IonQTranspileLevelWarning` filter, plus kept the
  `SparseEfficiencyWarning` ignore the PR branch had added. Non-
  overlapping union; no rules dropped from either side.
